File: cpp/algorithms/string/z_algorithm.cpp
// Z 函数（扩展 KMP）
#include <bits/stdc++.h>
#include <cassert>
using namespace std;

vector<int> z_function(const string& s) {
    int n = (int)s.size();
    vector<int> z(n, 0);
    int l = 0, r = 0;
    for (int i = 1; i < n; ++i) {
        if (i <= r) z[i] = min(r - i + 1, z[i - l]);
        while (i + z[i] < n && s[z[i]] == s[i + z[i]]) ++z[i];
        if (i + z[i] - 1 > r) l = i, r = i + z[i] - 1;
    }
    return z;
}

vector<int> z_search(const string& text, const string& pat) {
    vector<int> res;
    if (pat.empty()) return res;
    if ((int)pat.size() > (int)text.size()) return res;
    string combined = pat + '\0' + text;
    vector<int> z = z_function(combined);
    int m = (int)pat.size();
    for (int i = m + 1; i < (int)combined.size(); ++i) {
        if (z[i] == m) res.push_back(i - m - 1);
    }
    return res;
}

int main() {
    assert(z_function("aaabaab") == vector<int>({0, 2, 1, 0, 2, 1, 0}));
    assert(z_search("ababa", "aba") == vector<int>({0, 2}));
    cout << "z_algorithm OK" << endl;
    return 0;
}

File: python/algorithms/string/z_algorithm.py
"""Z 函数（扩展 KMP）：线性求每个后缀与前缀的最长公共前缀长度。"""

from __future__ import annotations


def z_function(s: str) -> list[int]:
    n = len(s)
    if n == 0:
        return []
    z = [0] * n
    l = r = 0
    for i in range(1, n):
        if i <= r:
            z[i] = min(r - i + 1, z[i - l])
        while i + z[i] < n and s[z[i]] == s[i + z[i]]:
            z[i] += 1
        if i + z[i] - 1 > r:
            l, r = i, i + z[i] - 1
    return z


def z_search(text: str, pat: str) -> list[int]:
    """返回 pat 在 text 中所有起始下标。"""
    if not pat:
        return list(range(len(text) + 1))
    if len(pat) > len(text):
        return []
    combined = pat + "\x00" + text
    z = z_function(combined)
    m = len(pat)
    return [i - m - 1 for i in range(m + 1, len(combined)) if z[i] == m]


if __name__ == "__main__":
    assert z_function("aaabaab") == [0, 2, 1, 0, 2, 1, 0]
    assert z_search("ababa", "aba") == [0, 2]
    assert z_function("") == []
    assert z_search("ab", "xyz") == []
    assert z_search("ab", "") == [0, 1, 2]
    print("z_algorithm OK")
